### PR TITLE
ReDoS: convert RelevantState to a class in the PrefixConstruction module

### PR DESCRIPTION
**Pull request by erik-krogh**
I just learned that `class X instanceof Y {` inside a parameterized module is allowed.  
So now I'm using it.  

Evaluations were uneventful: [Ruby](https://github.com/github/codeql-dca-main/tree/data/erik-krogh/paramClass-rb/reports), [Java](https://github.com/github/codeql-dca-main/tree/data/erik-krogh/paramClass-java__1/reports), [Python](https://github.com/github/codeql-dca-main/tree/data/erik-krogh/paramClass-py/reports), [JavaScript](https://github.com/github/codeql-dca-main/tree/data/erik-krogh/paramClass-js/reports).

### CODE REVIEW

File: java/ql/lib/semmle/code/java/security/regexp/NfaUtils.qll [hunk 1]
```diff
@@ -887,11 +887,10 @@ module PrefixConstruction<isCandidateSig/1 isCandidate> {
   /**
    * Holds if `state` is the textually last start state for the regular expression.
    */
-  private predicate lastStartState(State state) {
+  private predicate lastStartState(RelevantState state) {
     exists(RegExpRoot root |
       state =
         max(State s, Location l |
```
Comment (aibaars) [inline java/ql/lib/semmle/code/java/security/regexp/NfaUtils.qll:893]:
```suggestion
        max(RelevantState s, Location l |
```

**Review by aibaars — COMMENTED** [review]
Looks good to me. One suggestion.

**Review by yoff — DISMISSED** [review]
Sweet, this cleans up nicely :-)

**Comment by erik-krogh** [comment]
The QL-for-QL warnings can be ignored, they will be fixed by this PR: https://github.com/github/codeql/pull/10250

**Review by yoff — APPROVED** [review]
LGTM

**Pull request merged**